src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -0,0 +1,1341 @@
+            comp.VerifyDiagnostics(
+                // (6,9): warning CS8602: Possible dereference of a null reference.
+                //         F(default(object)).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(default(object))").WithLocation(6, 9),
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         F(default(string)).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(default(string))").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void TypeInference_Tuple()
+        {
+            var source =
+@"class C
+{
+    static (T, U) F<T, U>((T, U) t) => t;
+    static void G(string x, string? y)
+    {
+        var t = (x, y);
+        F(t).Item1.ToString();
+        F(t).Item2.ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         F(t).Item2.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(t).Item2").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void TypeInference_ObjectCreation()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G()
+    {
+        F(new C { }).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TypeInference_BinaryOperator()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G(string x, string? y)
+    {
+        F(x + x).ToString();
+        F(x + y).ToString();
+        F(y + x).ToString();
+        F(y + y).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TypeInference_NullCoalescingOperator()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G(object x, object? y)
+    {
+        F(x ?? x).ToString();
+        F(x ?? y).ToString();
+        F(y ?? x).ToString();
+        F(y ?? y).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (6,11): hidden CS8607: Expression is probably never null.
+                //         F(x ?? x).ToString();
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "x").WithLocation(6, 11),
```

Comment:
If we warn that it is *probably* never null, maybe we should also warn about possible dereference?